### DISCUSSION TITLE
umi中是否有提供获取项目根目录的相关方法

### DISCUSSION THREAD

**Discussion by wanghailong66** [umijs/umi#12067] [Q&A]
现状：
 我的项目是基于umi开发的，这个项目会发布到内网的npm私服供其他人引入使用，项目中关于网络请求的模块需要获取到使用者配置一下api请求的路径，因为umi是约定式开发（例如根目录下config/config.js/ts），可能我的项目就是需要获取根目录下config/api.js/ts来读取配置，所以就想问一下有没有相关工具类，或者实现思路~

待发布的项目：
<img width="903" alt="77051cfa3893777248d1202aadb5995" src="https://github.com/umijs/umi/assets/81939984/7ba000a9-50ef-47b4-9d4e-6277c70250e6">

使用者项目：
<img width="982" alt="835a375a855eeb39ec7419eeedddb05" src="https://github.com/umijs/umi/assets/81939984/6a7fbdeb-2c77-4bce-aa18-c4b69a6c69dd">

**Comment by fz6m** [comment]
发布到 npm 作为一个包来使用的时候，必须保证这个包是纯净的、可共用、通用的、任何人都可以使用的代码。

而 umi 是一个项目，包含了一些约定和自带的行为，这些行为不适用于放到第三方包里，任何项目（比如 umi 、vite、nextjs 等）也是不可以发布到 npm 作为包的。

如果你希望通用某些代码，可以把这些通用的代码抽取出来，然后发布为一个包，之后在需要用到的项目里安装然后导入使用就可以了，包只共享的是代码而不是约定。

**Reply by wanghailong66** [reply]
哇，感谢回复~

因为这个本身项目是低代码平台项目基盘，发布到npm其实是为了屏蔽源码和方便引入，但是其中需要开放一些口让开发人员配置以及写自定义代码的地方，所以才会有这个需求。
